### PR TITLE
add convert-gptq-to-ggml.py to bigdl-llama

### PR DESCRIPTION
**Pull request by qiuxin2012**
## Description

add python script to convert gptq quantized 4bits llama model to ggml type.

### 1. Why the change?

add python script to convert gptq quantized 4bits llama  model to ggml type.
Usage:

### 2. User API changes

a new python script to convert gptq 4bits llama model to ggml type.
Usage:
```
convert-gptq-to-ggml.py llamaXXb-4bit.pt tokenizer.model out.bin
```
For example:
```
python convert-gptq-to-ggml.py vicuna-13B-1.1-GPTQ-4bit-128g/vicuna-13B-1.1-GPTQ-4bit-128g.compat.no-act-order.pt vicuna-13B-1.1-GPTQ-4bit-128g/tokenizer.model vicuna-13B-1.1-gptq-4bit-ggml.bin
```

### 3. Summary of the change 

add python script convert-gptq-to-ggml.py
The script provide a way to convert gptq quantized 4bits llama model to ggml type. But don't support act-order and safetensors.

### 4. How to test?
- [ ] N/A
- [ ] Unit test
- [ ] Application test
- [ ] Document test
- [ ] ...

### CODE REVIEW

**Comment by qiuxin2012** [comment]
TODO: add windows support

**Comment by jason-dai** [comment]
I think we may need two command line tools for model conversion with the same arguments:
- `convert_pytorch_model` (or `convert_hf_model`?)
- `convert_gptq_model`

Or one `convert_model` command line tool to support both format? And where should we put the model conversion tool? @TheaperDeng

**Comment by jason-dai** [comment]
> ### 2. User API changes
> a new python script

Please describe how it is used

> 
> ### 3. Summary of the change
> add python script convert-gptq-to-ggml.py The script provide a way to convert gptq quantized 4bits model to ggml type. But don't support act-order and safetensors.

Can you check if the input format is expected and report error if not?

**Comment by jason-dai** [comment]
> > I think we may need two command line tools for model conversion with the same arguments:
> > 
> > * `convert_pytorch_model` (or `convert_hf_model`?)
> > * `convert_gptq_model`
> > 
> > Or one `convert_model` command line tool to support both format? And where should we put the model conversion tool? @TheaperDeng
> 
> We have a `convert_model` entry point now, I believe put it there with some extra parameters(args) should be suitable

What do you mean by `convert_model` entry point? Is it a script?

**Comment by jason-dai** [comment]
> > I think we may need two command line tools for model conversion with the same arguments:
> > 
> > * `convert_pytorch_model` (or `convert_hf_model`?)
> > * `convert_gptq_model`
> > 
> > Or one `convert_model` command line tool to support both format? And where should we put the model conversion tool? @TheaperDeng
> 
> We have a `convert_model` entry point now, I believe put it there with some extra parameters(args) should be suitable

What do you mean by `convert_model` entry point? Is it a script?

File: python/llm/src/bigdl/llm/gptq/convert/convert-gptq-to-ggml.py [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+import sys
+import json
+import struct
+import numpy as np
+import torch
+from sentencepiece import SentencePieceProcessor
+
+if len(sys.argv) != 4:
+    print("Usage: convert-gptq-to-ggml.py llamaXXb-4bit.pt tokenizer.model out.bin\n")
+    sys.exit(1)
```
Comment (jason-dai) [inline python/llm/src/bigdl/llm/gptq/convert/convert-gptq-to-ggml.py:32]: Is this for llama only?

**Comment by TheaperDeng** [comment]
> > > I think we may need two command line tools for model conversion with the same arguments:
> > > 
> > > * `convert_pytorch_model` (or `convert_hf_model`?)
> > > * `convert_gptq_model`
> > > 
> > > Or one `convert_model` command line tool to support both format? And where should we put the model conversion tool? @TheaperDeng
> > 
> > 
> > We have a `convert_model` entry point now, I believe put it there with some extra parameters(args) should be suitable
> 
> What do you mean by `convert_model` entry point? Is it a script?

Yes, it's a python script but could work like a cmd line tool:
https://github.com/intel-analytics/BigDL/blob/main/python/llm/setup.py#L179

It work like this:
https://github.com/TheaperDeng/BigDL/blob/85d6b243bea829b7f7c55c5dd723f8e5b9cecb3d/python/llm/README.md#convert-your-model

File: python/llm/src/bigdl/llm/gptq/convert/convert-gptq-to-ggml.py [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+        for j in range(i, i + (32 // bits)):
+            zeros[:, j] = (qzeros[:, col] >> (bits * (j - i)) & (2 ** bits - 1)) + 1
+        i += 32 // bits
+        col += 1
+    return zeros
+
+def convert_q4_with_groupsize(src_name, dst_name, permute=False):
+    qzeros = model[f"{src_name}.qzeros"].numpy()
+    zeros = qzeros_to_zeros(qzeros).T
+    scales = model[f"{src_name}.scales"].numpy().T
```
Comment (jason-dai) [inline python/llm/src/bigdl/llm/gptq/convert/convert-gptq-to-ggml.py:149]: Are we using fp16 or fp32 scale/zero? I think the latest llama.cpp use fp16 by default, and we have changed to fp32 in our implementation for q4_0. @MeouSker77

Comment (MeouSker77) [reply python/llm/src/bigdl/llm/gptq/convert/convert-gptq-to-ggml.py:149]: yes, we have changed to fp32 for q4_0

Comment (jason-dai) [reply python/llm/src/bigdl/llm/gptq/convert/convert-gptq-to-ggml.py:152]: we should probably test the latency & accuracy of fp32 vs. fp16

**Comment by shane-huang** [comment]
IMHO, using one command `convert_model` with options like `pytorch` or `gptq` is better, as later we may support more model format, or if the format name is changed we don't have to change the command name. Btw, can we infer the format based on the model suffix or something? 

Another thing, shall we rename the command for conversion and running to `bigdl-llm-convert`, and `bigdl-llm`? That'll be easier to search for.

**Comment by jason-dai** [comment]
> IMHO, using one command `convert_model` with options like `pytorch` or `gptq` is better, as we may support more model formats in future. Btw, can we infer the format based on the model suffix or something?
> 
> Another thing, shall we rename the command for conversion and running to `bigdl-llm-convert`, and `bigdl-llm`? That'll be easier to search for. Or maybe just using one command `bigdl-llm-cli` and use `bidl-llm-cli convert` for conversion and other options compatible with llama-cpp cli.
> 
> @jason-dai @TheaperDeng

I don't think one can easily infer the model format or family.

And it's probably simpler in concept to have a separate model conversion tool; we can call it `bigdl-llm-convert` but maybe the name is a bit long. Or we can just use `llm-convert` (and similar `llm-cli`)?

File: python/llm/src/bigdl/llm/gptq/convert/convert-gptq-to-ggml.py (hunk 1 above)
Comment (qiuxin2012) [reply python/llm/src/bigdl/llm/gptq/convert/convert-gptq-to-ggml.py]: yes

Comment (jason-dai) [reply python/llm/src/bigdl/llm/gptq/convert/convert-gptq-to-ggml.py]: The we should ask the user always specify `model_family=llama` in the command line

**Comment by qiuxin2012** [comment]
`:=` is new from python3.8, our style check is based on 3.7. So need to remove `:=` to solve the style check.
